File: dart/lib/parrot.dart
import 'dart:math';

import 'package:parrot_refactoring_kata/parrot_type.dart';

class Parrot {
  final ParrotType type;
  final int numberOfCoconuts;
  final double voltage;
  final bool isNailed;

  Parrot(
    this.type, {
    required this.numberOfCoconuts,
    required this.voltage,
    required this.isNailed,
  });

  double speed() {
    switch (type) {
      case ParrotType.european:
        return baseSpeed;

      case ParrotType.african:
        return max(0, baseSpeed - loadFactor * numberOfCoconuts);

      case ParrotType.norwegianBlue:
        return isNailed ? 0 : _baseSpeed(voltage);
    }
  }

  double _baseSpeed(double voltage) {
    return min(24.0, voltage * baseSpeed);
  }

  double get loadFactor => 9.0;

  double get baseSpeed => 12.0;
}

File: Elixir/lib/parrot.ex
defmodule Parrot do
  def speed(type, number_of_coconuts, voltage, is_nailed) do
    case type do
      :european ->
        base_speed()

      :african ->
        max(0, base_speed() - load_factor() * number_of_coconuts)

      :norwegian_blue ->
        if is_nailed, do: 0, else: base_speed_with_voltage(voltage)

      _ ->
        raise "Unknown parrot type"
    end
  end

  def cry(type, _number_of_coconuts, voltage, _is_nailed) do
    case type do
      :european ->
        "Sqoork!"

      :african ->
        "Sqaark!"

      :norwegian_blue ->
        if voltage > 0, do: "Bzzzzzz", else: "..."

      _ ->
        raise "Unknown parrot type"
    end
  end

  defp base_speed, do: 12.0

  defp load_factor, do: 9.0

  defp base_speed_with_voltage(voltage) do
    min(24.0, voltage * base_speed())
  end
end
